Define atomicity (#2107)
oasis-tcs/odata-specs: odata-protocol/11.4 Data Modification.md

A [success response](#SuccessResponses) indicates that data have been modified,
regardless of whether the requested content could be returned.

> Data modification requests guarantee _atomicity_ in the following sense:

Reviewer: perhaps have a separate subsection for atomicity for better referencing.

When the request completes successfully, every subsequent request observes a state of the system
in which either all or none of the changes have been carried out.
But depending on how a service makes this guarantee, clients MAY observe
a state of the system where the changes have been carried out only partially
while the data modification request is still being executed.

> A transactional service is a service that protects against seeing such partial changes.

Reviewer: Might have to have a better term than "Transactional" -- a transactional service may have different isolation levels.

> But for services that do not require this degree of transactional consistency,

Reviewer: ```suggestion
For services that do not require this degree of transactional consistency,
```

[Data Modification Requests](#DataModification) share the following
semantics.

#### ##subsubsubsec Atomicity

Data modification requests guarantee _atomicity_ in the following sense:
When the request completes, every subsequent request observes a state of the system
in which either all or none of the changes have been carried out.

> But depending on how a service makes this guarantee, clients MAY observe

Reviewer: ```suggestion
Depending on how a service makes this guarantee, clients MAY observe
```

> a state of the system where the changes have been carried out only partially
> while the data modification request is still being executed.

Reviewer: ```suggestion
a state of the system where the changes have been carried out only partially
while the data modification request is still being executed, including changes that may later be rolled back.
```

[Data Modification Requests](#DataModification) share the following
semantics.

#### ##subsubsubsec Atomicity

Data modification requests guarantee _atomicity_ in the following sense:
When the request completes, every subsequent request observes a state of the system
in which either all or none of the changes have been carried out.
The experience of concurrent clients observing the state of the system
can be described through isolation levels as defined in [SQL92, section 4.28](#_SQL92),
even for services that do not use a transactional database.

When a service operates at isolation level "read uncommitted",
concurrent clients MAY observe a state of the system where the changes have been carried out only partially
while the data modification request is still being executed.

> During the execution, the service MUST roll back any changes that
> may have been applied (and seen by clients) before another change failed.

Reviewer: ```suggestion
In such a case, if an error is encountered while processing the request, the service MUST roll back any changes that
have been applied (even though they may have been seen by clients).
```

When a service operates at an isolation level of "read committed" or higher,
it protects against seeing such partial changes.

> When data modification requests apply the
> [`continue-on-error`](#Preferencecontinueonerrorodatacontinueonerror) preference,
> they do not always guarantee atomicity. See the sections below where this preference is mentioned.

Reviewer: ```suggestion
When data modification requests apply the
[`continue-on-error`](#Preferencecontinueonerrorodatacontinueonerror) preference,
they do not guarantee atomicity. See the sections below where this preference is mentioned.
```

annotated with the term [`Core.DataModificationException`]($$$OData-VocCore$$$#DataModificationException) and SHOULD specify
a `responseCode` of `424` ([Failed Dependency](#ResponseCode424FailedDependency)).

> Regardless of the `continue-on-error` preference, the collection update must happen

Reviewer: Continue-on-error necessarily means non-atomic.

Reviewer: In the absence of continue-on-error, the operation must complete in an atomic manner.
